hazelcast-client-legacy/src/test/java/com/hazelcast/client/ClientReconnectTest.java

```diff
@@ -69,4 +77,34 @@ public void stateChanged(LifecycleEvent event) {
         assertEquals("test", m.get("test"));
     }
 
+    @Test
+    public void testReconnectToNewInstanceAtSameAddress() throws InterruptedException {
+        HazelcastInstance instance = hazelcastFactory.newHazelcastInstance();
+        Address localAddress = instance.getCluster().getLocalMember().getAddress();
+        final HazelcastInstance client = hazelcastFactory.newHazelcastClient();
+
+        final CountDownLatch memberRemovedLatch = new CountDownLatch(1);
+        client.getCluster().addMembershipListener(new MembershipAdapter() {
+            @Override
+            public void memberRemoved(MembershipEvent membershipEvent) {
+                memberRemovedLatch.countDown();
+            }
+        });
+
+        instance.shutdown();
+        final HazelcastInstance instance2 = hazelcastFactory.newHazelcastInstance(localAddress);
+
+        assertTrue(memberRemovedLatch.await(30, TimeUnit.SECONDS));
```

Comment:
we have an assertOpensEventually just for this to prevent coming up with a period.